src/main/java/com/pingcap/tikv/meta/TiKey.java

```diff
@@ -0,0 +1,113 @@
+/*
+ *
+ * Copyright 2017 PingCAP, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.pingcap.tikv.meta;
+
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.collect.Range;
+import com.google.common.primitives.UnsignedBytes;
+import com.google.protobuf.ByteString;
+import com.pingcap.tikv.kvproto.Coprocessor;
+import java.util.Comparator;
+import javax.annotation.Nonnull;
+
+public class TiKey<T> implements Comparable<TiKey<T>> {
+
+  // below might uses UnsafeComparator if possible
+  private static final Comparator<byte[]> comparator = UnsignedBytes.lexicographicalComparator();
+  private T data;
+
+  public TiKey(@Nonnull T data) {
+    this.data = data;
+  }
+
+  private int compareTo(@Nonnull ByteString o) {
+    requireNonNull(o, "other is null");
+    ByteString data = (ByteString)this.data;
```

Comment:
maybe we should add <instanceof> if-clause to deal with exceptions?
Here we have dealt with TiKey's comparison against Comparable, ByteString and byte[], but not yet resolved cases vice versa. 
The same goes following two compareTo functions.